File: tests/cpp/unit/ProcessPriorityMonitorTests.cpp
#include "services/ProcessPriorityMonitor.h"
#include "services/IProcessPriorityService.h"
#include "models/ProcessPriorityRule.h"

#include <gtest/gtest.h>

#include <atomic>
#include <memory>
#include <unordered_map>

using namespace imeindicator;

namespace {

// 単純なフェイク実装（モック）。GoogleMock は別途リンク不要。
class FakePriorityService : public services::IProcessPriorityService {
public:
    std::vector<services::ProcessPriorityEntry>
        getProcessPriorities(const std::wstring& name) override
    {
        getCalls.fetch_add(1);
        auto it = mockProcesses.find(name);
        if (it == mockProcesses.end()) return {};
        return it->second;
    }
    bool setPriority(DWORD pid, DWORD pc) override
    {
        setCalls.fetch_add(1);
        if (failSetPriority) return false;
        for (auto& kv : mockProcesses) {
            for (auto& e : kv.second) {
                if (e.processId == pid) e.currentPriorityClass = pc;
            }
        }
        return true;
    }
    std::optional<DWORD_PTR> getAffinity(DWORD /*pid*/) override
    {
        return DWORD_PTR{1};
    }
    bool setAffinity(DWORD /*pid*/, DWORD_PTR) override { return true; }

    std::unordered_map<std::wstring, std::vector<services::ProcessPriorityEntry>> mockProcesses;
    std::atomic<int> getCalls{0};
    std::atomic<int> setCalls{0};
    bool failSetPriority{false};
};

models::ProcessPriorityRule makeRule(const wchar_t* name, models::PriorityLevel pri)
{
    models::ProcessPriorityRule r;
    r.processName = name;
    r.targetPriority = pri;
    r.maxBackoffExponent = 6;
    r.isEnabled = true;
    r.useECoreOnly = false;
    return r;
}

} // namespace

TEST(ProcessPriorityMonitorTests, EvaluateRuleSkippedWhenNoProcess)
{
    auto svc = std::make_shared<FakePriorityService>();
    services::ProcessPriorityMonitor mon(svc);
    auto rule = makeRule(L"chrome", models::PriorityLevel::BelowNormal);
    services::RuleRuntimeState state(rule, 1);
    mon.evaluateRule(rule, state);
    EXPECT_EQ(state.lastResult().value(), services::MonitorResult::Skipped);
}

TEST(ProcessPriorityMonitorTests, EvaluateRuleChangesPriority)
{
    auto svc = std::make_shared<FakePriorityService>();
    svc->mockProcesses[L"chrome"] = {{1234, NORMAL_PRIORITY_CLASS}};
    services::ProcessPriorityMonitor mon(svc);
    auto rule = makeRule(L"chrome", models::PriorityLevel::BelowNormal);
    services::RuleRuntimeState state(rule, 1);
    mon.evaluateRule(rule, state);
    EXPECT_EQ(state.lastResult().value(), services::MonitorResult::Success);
    EXPECT_GE(svc->setCalls.load(), 1);
}

TEST(ProcessPriorityMonitorTests, EvaluateRuleSkippedWhenAlreadyTarget)
{
    auto svc = std::make_shared<FakePriorityService>();
    svc->mockProcesses[L"chrome"] = {{1234, BELOW_NORMAL_PRIORITY_CLASS}};
    services::ProcessPriorityMonitor mon(svc);
    auto rule = makeRule(L"chrome", models::PriorityLevel::BelowNormal);
    services::RuleRuntimeState state(rule, 1);
    mon.evaluateRule(rule, state);
    EXPECT_EQ(state.lastResult().value(), services::MonitorResult::Skipped);
}

TEST(ProcessPriorityMonitorTests, EvaluateRuleFailedWhenSetFails)
{
    auto svc = std::make_shared<FakePriorityService>();
    svc->mockProcesses[L"chrome"] = {{1234, NORMAL_PRIORITY_CLASS}};
    svc->failSetPriority = true;
    services::ProcessPriorityMonitor mon(svc);
    auto rule = makeRule(L"chrome", models::PriorityLevel::BelowNormal);
    services::RuleRuntimeState state(rule, 1);
    mon.evaluateRule(rule, state);
    EXPECT_EQ(state.lastResult().value(), services::MonitorResult::Failed);
}

TEST(ProcessPriorityMonitorTests, UpdateRulesSyncsRuntimeStates)
{
    auto svc = std::make_shared<FakePriorityService>();
    services::ProcessPriorityMonitor mon(svc);
    std::vector<models::ProcessPriorityRule> rules = {
        makeRule(L"chrome", models::PriorityLevel::BelowNormal),
        makeRule(L"opera", models::PriorityLevel::Normal),
    };
    mon.updateRules(rules);
    EXPECT_EQ(mon.runtimeStatesCount(), 2u);
    rules.pop_back();
    mon.updateRules(rules);
    EXPECT_EQ(mon.runtimeStatesCount(), 1u);
    mon.stop();
}

File: tests/csharp/IMEIndicator.Tests/Services/ProcessPriorityMonitorTests.cs
// Copyright (C) 2026 IMEIndicator Project
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License v2 or later.
// See COPYING in the repository root for the full license text.

using System.Diagnostics;
using IMEIndicator.Models;
using IMEIndicator.Services;
using Xunit;

namespace IMEIndicator.Tests.Services;

/// <summary>
/// <see cref="ProcessPriorityMonitor"/> のテスト。現行 tests/cpp/unit/ProcessPriorityMonitorTests.cpp の移植。
/// </summary>
public sealed class ProcessPriorityMonitorTests
{
    // 現行 FakePriorityService（C++）と等価のフェイク実装。
    private sealed class FakePriorityService : IProcessPriorityService
    {
        public Dictionary<string, List<ProcessPriorityEntry>> MockProcesses { get; } = new(StringComparer.OrdinalIgnoreCase);
        public int GetCalls;
        public int SetCalls;
        public bool FailSetPriority;

        public IReadOnlyList<ProcessPriorityEntry> GetProcessPriorities(string processNameNoExt)
        {
            Interlocked.Increment(ref GetCalls);
            // 独立したコピーを返す（C++ 版の std::vector 値返しと同じ意味。呼び出し元が列挙している間に
            // SetPriority が内部状態を書き換えても、既に取得済みのスナップショットには影響しない）。
            return MockProcesses.TryGetValue(processNameNoExt, out List<ProcessPriorityEntry>? list) ? [.. list] : [];
        }

        public bool SetPriority(int processId, ProcessPriorityClass priorityClass)
        {
            Interlocked.Increment(ref SetCalls);
            if (FailSetPriority)
            {
                return false;
            }
            foreach (List<ProcessPriorityEntry> list in MockProcesses.Values)
            {
                for (int i = 0; i < list.Count; i++)
                {
                    if (list[i].ProcessId == processId)
                    {
                        list[i] = list[i] with { CurrentPriorityClass = priorityClass };
                    }
                }
            }
            return true;
        }

        public nint? GetAffinity(int processId) => 1;

        public bool SetAffinity(int processId, nint affinityMask) => true;

        public bool IsAccessibleForControl(string processNameNoExt) => true;
    }

    private static ProcessPriorityRule MakeRule(string name, PriorityLevel priority) => new()
    {
        ProcessName = name,
        TargetPriority = priority,
        MaxBackoffExponent = 6,
        IsEnabled = true,
        UseECoreOnly = false,
    };

    [Fact]
    public void EvaluateRule_SkippedWhenNoProcess()
    {
        var svc = new FakePriorityService();
        var mon = new ProcessPriorityMonitor(svc);
        ProcessPriorityRule rule = MakeRule("chrome", PriorityLevel.BelowNormal);
        var state = new RuleRuntimeState(rule, 1);

        mon.EvaluateRule(rule, state);

        Assert.Equal(MonitorResult.Skipped, state.LastResult);
    }

    [Fact]
    public void EvaluateRule_ChangesPriority()
    {
        var svc = new FakePriorityService();
        svc.MockProcesses["chrome"] = [new ProcessPriorityEntry(1234, ProcessPriorityClass.Normal)];
        var mon = new ProcessPriorityMonitor(svc);
        ProcessPriorityRule rule = MakeRule("chrome", PriorityLevel.BelowNormal);
        var state = new RuleRuntimeState(rule, 1);

        mon.EvaluateRule(rule, state);

        Assert.Equal(MonitorResult.Success, state.LastResult);
        Assert.True(svc.SetCalls >= 1);
    }

    [Fact]
    public void EvaluateRule_SkippedWhenAlreadyTarget()
    {
        var svc = new FakePriorityService();
        svc.MockProcesses["chrome"] = [new ProcessPriorityEntry(1234, ProcessPriorityClass.BelowNormal)];
        var mon = new ProcessPriorityMonitor(svc);
        ProcessPriorityRule rule = MakeRule("chrome", PriorityLevel.BelowNormal);
        var state = new RuleRuntimeState(rule, 1);

        mon.EvaluateRule(rule, state);

        Assert.Equal(MonitorResult.Skipped, state.LastResult);
    }

    [Fact]
    public void EvaluateRule_FailedWhenSetFails()
    {
        var svc = new FakePriorityService();
        svc.MockProcesses["chrome"] = [new ProcessPriorityEntry(1234, ProcessPriorityClass.Normal)];
        svc.FailSetPriority = true;
        var mon = new ProcessPriorityMonitor(svc);
        ProcessPriorityRule rule = MakeRule("chrome", PriorityLevel.BelowNormal);
        var state = new RuleRuntimeState(rule, 1);

        mon.EvaluateRule(rule, state);

        Assert.Equal(MonitorResult.Failed, state.LastResult);
    }

    [Fact]
    public void UpdateRules_SyncsRuntimeStates()
    {
        var svc = new FakePriorityService();
        var mon = new ProcessPriorityMonitor(svc);
        List<ProcessPriorityRule> rules =
        [
            MakeRule("chrome", PriorityLevel.BelowNormal),
            MakeRule("opera", PriorityLevel.Normal),
        ];

        mon.UpdateRules(rules);
        Assert.Equal(2, mon.RuntimeStatesCount);

        rules.RemoveAt(rules.Count - 1);
        mon.UpdateRules(rules);
        Assert.Equal(1, mon.RuntimeStatesCount);

        mon.Stop();
    }

    [Fact]
    public void EvaluateRule_DisabledRule_IsSkipped()
    {
        var svc = new FakePriorityService();
        svc.MockProcesses["chrome"] = [new ProcessPriorityEntry(1234, ProcessPriorityClass.Normal)];
        var mon = new ProcessPriorityMonitor(svc);
        ProcessPriorityRule rule = MakeRule("chrome", PriorityLevel.BelowNormal);
        rule.IsEnabled = false;
        var state = new RuleRuntimeState(rule, 1);

        mon.EvaluateRule(rule, state);

        Assert.Equal(MonitorResult.Skipped, state.LastResult);
        Assert.Equal(0, svc.GetCalls);
    }

    [Fact]
    public void StartAndStop_DoesNotThrow()
    {
        var svc = new FakePriorityService();
        using var mon = new ProcessPriorityMonitor(svc);
        List<ProcessPriorityRule> rules = [MakeRule("chrome", PriorityLevel.BelowNormal)];

        mon.Start(rules, 1);
        Assert.Equal(1, mon.RuntimeStatesCount);
        mon.Stop();
    }
}
